src/views/Browse/index.tsx

```diff
@@ -1,7 +1,59 @@
-import { useLocation } from "react-router-dom";
+import { useEffect, useMemo, useState } from "react";
+import { useLocation, useNavigate } from "react-router";
+
+import { ApiBrowseRes } from "../../common/types";
+import { useSmartFetch } from "../../hooks";
+import { useToasts } from "../../providers";
+import { Browser } from "./Browser";
+import { FilterProvider, SortProvider } from "./providers";
 
 export const Browse: React.FC = () => {
   const location = useLocation();
+  const navigate = useNavigate();
+  const { smartFetch, notifyUnknown } = useSmartFetch();
+  const { toast } = useToasts();
+
+  // TODO: is there a better way?
+  const path = useMemo(() => location.pathname.replace(/^\/browse/, ""), [location.pathname]);
+
+  const [isLoading, setIsLoading] = useState(false);
+  const [contents, setContents] = useState<ApiBrowseRes | undefined>();
+
+  useEffect(() => {
+    (async () => {
+      setIsLoading(true);
+
+      const result = await smartFetch<ApiBrowseRes>(
+        `/api/browse?path=${path}`,
+        { method: "GET" },
+        { handleUnknown: false },
+      );
+      if (result.status !== "stale") setIsLoading(false);
+
+      switch (result.status) {
+        case "ok":
+          setContents(result.parsed);
+          break;
+        case "unknown":
+          if (result.response.status === 404) {
+            toast(
+              `We couldn't find the place you were trying to browse ('${path}'). We've taken you back to the safety of the harbour ⚓`,
+              "error",
+              { id: "navigate", hideIn: 0 },
+            );
+            navigate("/browse");
+          } else {
+            notifyUnknown();
+          }
+      }
+    })();
+  }, [smartFetch, path, toast, notifyUnknown, navigate]);
```

Comment:
I think there is a race condition where displayed content is not the one last requested (when a user navigates back and forth between directories without waiting for request to finish)

Reply:
I've tested just to make sure I'm not missing something obvious, but there shouldn't be. We're using [`smartFetch` which checks whether the response it receives is to the latest request that has been sent](https://github.com/tadeaspetak/teleport-challenge/blob/1df68dabeac2553e958aff42b469b49f3040820a/src/hooks/useSmartFetch.ts#L58). When it's not, it returns `status: stale`. The contents of the `Browser` are only updated when the result has `status: ok`.

Or is there another race condition I've missed?

Reply:
i've missed that out thanks for clarifying it.